### PR TITLE
setup.py: Recommend installation command for pkgs

### PR DESCRIPTION
**Pull request by AbdealiJK**
If a package is not found, provide a better error message
that also explains how to install the package or gives the
user a link explaining what to do to install the package.

The supported methods are:
 - Provide a link for windows
 - Provide a command using a package manager for the OS
   (For example: apt-get for Ubuntu/Debian, dnf and yum for
    Fedora/CentOS/RHEL, brew and port for OSX)

Fixes https://github.com/matplotlib/matplotlib/issues/6546

### CODE REVIEW

**Comment by AbdealiJK** [comment]
This is still a work in progress, I haven't tested it to my satisfaction - but would love some feedback on the method used as I've never contributed to matplotlib before.

File: setupext.py [hunk 1]
```diff
@@ -531,6 +540,38 @@ def do_custom_build(self):
...
+        """
+        The help message to show if the package is not installed. The help
+        message shown depends on whether some class variables are present.
+        """
+        def _try_managers(*managers):
+            for manager in managers:
+                pkg_name = self.pkg_names.get(manager, None)
+                if pkg_name:
+                    try:
+                        _ = check_output(["which", manager],
```
Comment (QuLogic) [inline setupext.py]: There is a [`shutil.which`](https://docs.python.org/3/library/shutil.html), but unfortunately, it's 3.3+. It's probably not worth it to conditionally call it, but can you add a comment about switching to it when we drop 2.7?

File: setupext.py [hunk 2]
```diff
@@ -531,6 +540,38 @@ def do_custom_build(self):
...
+        if sys.platform == "win32":
+            url = self.pkg_names.get("windows_url", None)
+            if url:
+                message = ('Please check {0} for instructions to install {1}'
+                           .format(url, self.name))
+        elif sys.platform == "darwin":
+            message = _try_managers("brew", "port")
+        elif sys.platform.startswith("linux"):
+            release = platform.linux_distribution()[0].lower()
+            if release in ('debian', 'ubuntu'):
```
Comment (QuLogic) [inline setupext.py]: What does this do on Debian/Ubuntu derivatives like Linux Mint?

File: setupext.py [hunk 3]
```diff
@@ -531,6 +540,38 @@ def do_custom_build(self):
...
+                message = _try_managers('apt-get')
+            elif release in ('centos', 'redhat', 'fedora'):
+                message = _try_managers('dnf', 'yum')
+        return message
```
Comment (QuLogic) [inline setupext.py]: Need two blank lines between classes.

File: setupext.py [hunk 4]
```diff
@@ -531,6 +540,41 @@ def do_custom_build(self):
...
+        message = None
+        if sys.platform == "win32":
+            url = self.pkg_names.get("windows_url", None)
+            if url:
+                message = ('Please check {0} for instructions to install {1}'
+                           .format(url, self.name))
+        elif sys.platform == "darwin":
+            message = _try_managers("brew", "port")
+        elif sys.platform.startswith("linux"):
+            release = platform.linux_distribution()[0].lower()
```
Comment (AbdealiJK) [inline setupext.py]: I had a note I wanted to mention here. `platform.linux_distribution()` is deprecated in Py 3.5 and is planned to be removed in Py 3.7. The related bug can be found at https://bugs.python.org/issue1322

I found the package https://pypi.python.org/pypi/distro which makes a pypi package out of this, but I am not sure of a good way to install this to use it in the setup.py. Suggestions ? (`setup_requires` is meant for this, but it's known to be buggy IIRC)

File: setupext.py (hunk 2 above)
Comment (AbdealiJK) [inline setupext.py]: The supported distros in that function are:
```

_supported_dists = (
    'SuSE', 'debian', 'fedora', 'redhat', 'centos',
    'mandrake', 'mandriva', 'rocks', 'slackware', 'yellowdog', 'gentoo',
    'UnitedLinux', 'turbolinux')
```
If this is run on a non-supported distro, it returns empty strings `("", "", "")`.

There are also some issues with Ubuntu btw ... Some Ubuntu systems are detected as "debian" - http://bugs.python.org/issue9514

**Comment by AbdealiJK** [comment]
Bump - reminder for review of this PR

File: setupext.py (hunk 4 above)
Comment (tacaswell) [inline setupext.py]: That is a 'fun' bug report.

File: setupext.py [hunk 5]
```diff
@@ -531,6 +540,41 @@ def do_custom_build(self):
         """
         pass
 
+    def install_help_msg(self):
+        """
+        The help message to show if the package is not installed. The help
+        message shown depends on whether some class variables are present.
```
Comment (tacaswell) [inline setupext.py]: Can you add a note to this docstring explaining how to add the feature to sub-classes?

Comment (AbdealiJK) [inline setupext.py]: done

**Comment by tacaswell** [comment]
:+1: I'll merge this as soon as CI passes again (just to be paranoid).

**Comment by AbdealiJK** [comment]
@tacaswell Do you recommend we use `distro` instead of `platform.linux_distribution` as users will get a DeprecationWarning when they run this on py3.5 Similar issue: https://github.com/pypa/pip/issues/3823

**Comment by tacaswell** [comment]
It will only be seen if things go wrong _anyway_ :smiling_imp: .  

Given the amount of sound and fury in the upstream bug, I am inclined to let this shake it self out a bit more before jumping on a solution.

On the other hand, better to deal with it now than later (when we have forgotten).

**Comment by AbdealiJK** [comment]
Maybe we could merge the PR but keep the issue open or create a new tracking issue to keep track of it ?

**Pull request merged**